codalab/worker/worker.py

```diff
@@ -476,6 +476,7 @@ def process_runs(self):
         for uuid in self.runs:
             run_state = self.runs[uuid]
             self.runs[uuid] = self.run_state_manager.transition(run_state)
+            self.save_time_stats(uuid, run_state)
```

Comment:
I think we should inline the function `save_time_stats` directly in this function, so it's clearer what is actually going on. It's not clear that  `save_time_stats` not only ends the current stage, but starts measuring the next stage, so it would be clearer to inline.

```
        self.end_stage(uuid, from_stage.stage)
        self.start_stage_stats(uuid, self.runs[uuid].stage)
```

Moreover, there's an edge case that this code doesn't apply to. It's possible for `self.run_state_manager.transition` to return the same stage without changing it. In that case, we don't want to reset the stats for the stage, so maybe we should add some code like this:

```
            prev_state = self.runs[uuid]
            self.runs[uuid] = self.run_state_manager.transition(prev_state)
            if prev_state.stage != self.runs[uuid].stage:
                 self.end_stage(uuid, prev_state.stage)
                 self.start_stage_stats(uuid, self.runs[uuid].stage)
```

Reply:
makes sense to me. thanks for the detailed feedback!